<a href="https://colab.research.google.com/github/flzhfi/deep-learning-assignments/blob/main/week04-keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**[4주차]과제**
- 아래의 과제1), 과제2)의 코드를 완성하시오.
- 모든 코드의 결과를 출력하여, .ipynb의 링크를 **[4주차]/[4주차]과제**에 제출하시오.\
(실습 제출 예시: 4주차_2020XXXX_이름.ipynb 코드 링크)

In [ ]:
# 학번, 이름 출력
print(f'학번 : 2353764 : 신예진')

학번 : 2353764 : 신예진


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# google drive 연결
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# 구글 드라이브에 데이터를 넣고 path 연결 필요
data_path = '/content/drive/MyDrive/deeplearning/week04-test'

## 과제 1
- 케라스를 이용하여 은닉층이 2개인 MLP를 생성하고 옵티마이저를 Adam, RMSprop, Adadelta, Adagrad로 변경해보면서 성능을 측정한 뒤 어떤 방법이 가장 좋은 성능을 보이는지 확인하고, 각 방법에 대해서 서술하시오.
- 은닉층의 노드 수는 512, 256로 설정
- 은닉층의 Activation Function은 ReLU로 설정
- 모델 레이어 추가  
model.add(tf.keras.layers.Dense(노드갯수, activation='activation_func', kernel_initializer=initializer))  
<출력 예시>  
테스트 정확도 : 0.2779000282287598


In [ ]:
# 런타임 재시작 필요
!pip install scikeras==0.13.0 scikit-learn==1.4.2

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 91.7 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
umap-learn 0.5.11 requires scikit-learn>=1.6, but you have scikit-learn 1.4.2 which is incompatible.
hdbscan 0.8.41 requires scikit-learn>=1.6, but you have scikit-learn 1.4.2 which is incompatible.


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from sklearn.model_selection import GridSearchCV
from scikeras.wrappers import KerasClassifier, KerasRegressor
import os
import numpy as np

In [ ]:
# 실험 결과 재현을 위한 seed 고정
SEED = 4
os.environ['PYTHONHASHSEED'] = str(SEED)
os.environ['TF_DETERMINISTIC_OPS'] = '1'
tf.random.set_seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
initializer = tf.keras.initializers.GlorotUniform(seed=SEED)

In [ ]:
# 데이터세트 준비
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


In [ ]:
# 데이터 전처리
train_images = train_images.reshape((60000, 784)) # 2D 이미지 -> 벡터로 평탄화
train_images = train_images.astype('float32') / 255 # 정규화

test_images = test_images.reshape((10000, 784))
test_images = test_images.astype('float32') / 255

train_labels = tf.keras.utils.to_categorical(train_labels)
test_labels = tf.keras.utils.to_categorical(test_labels)

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from sklearn.model_selection import GridSearchCV
from scikeras.wrappers import KerasClassifier, KerasRegressor
import os
import numpy as np

# 신경망 모델 구축
def build_model():
    network = tf.keras.models.Sequential()

    # 은닉층 생성
    # 여기에 코드를 작성해주세요
    network.add(tf.keras.layers.Dense(512, activation='relu', input_shape=(28*28,),
                                      kernel_initializer=initializer))
    network.add(tf.keras.layers.Dense(256, activation='relu',
                                      kernel_initializer=initializer))

    network.add(tf.keras.layers.Dense(10, activation='sigmoid',
                                      kernel_initializer=initializer))

    # 옵티마이저 설정 필요
    network.compile(optimizer='Adam', # Adam 사용
                   loss='mse',
                   metrics=['accuracy'])
    return network

In [ ]:
# model 생성
model = build_model()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
result = model.fit(train_images, train_labels, epochs=5, batch_size=64)
# epochs = 5
# batch_size = 64: 64개씩 묶어서

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 11s 11ms/step - accuracy: 0.9118 - loss: 0.0139
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 9s 10ms/step - accuracy: 0.9748 - loss: 0.0043
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 11s 11ms/step - accuracy: 0.9839 - loss: 0.0029
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 13s 14ms/step - accuracy: 0.9885 - loss: 0.0022
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 8s 9ms/step - accuracy: 0.9905 - loss: 0.0017


In [ ]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 12ms/step - accuracy: 0.9819 - loss: 0.0029


In [ ]:
print('테스트 정확도 : ', test_acc)

테스트 정확도 :  0.9818999767303467


### 각 옵티마이저의 성능 작성
- RMSprop : 정확도(0.9761000275611877), loss 수렴속도(0.0041)
- Adagrad : 정확도(0.439300000667572), loss 수렴속도(0.0864)
- Adadelta : 정확도(0.14300000667572021), loss 수렴속도(0.0990)
- Adam : 정확도(0.9818999767303467), loss 수렴속도(0.0029)

### 가장 좋은 성능의 옵티마이저 작성
- 가장 좋은 성능: Adam

### 옵티마이저 서술
- RMSprop : 정확도가 Adam 다음으로 높게 나옴. 학습률이 0으로 떨어지지 않음.

- Adagrad : 정확도가 떨어짐. 학습이 진행될수록 학습률을 계속 줄이는 특성이 있음. 초반에는 정확도가 오르다가 중간에 정체되어 하락함.

- Adadelta : 학습 속도가 느리고, 성능이 좋지 못함. 학습률 파라미터가 아예 없음.

- Adam : 가장 빠르고 정확함. 현재 데이터셋에 적합하여 정확도가 제일 높게 나옴.

## 과제 2 MLP 모델의 정확도(Accuracy)를 71이상 달성하시오.  
- 해당 diabetes.csv 데이터는 Pima 인디언의 의료 기록과 각 환자가 5년 이내에 당뇨병 발병 여부가 저장되어 있다.
- 현재 모델은 1 : 당뇨병 양성, 0 : 당뇨병 음성을 분류하는 MLP 모델이다.
- 해당 모델의 하이퍼 파라미터를 수정하여 정확도 71 이상을 달성하여라.
- 하이퍼 파라미터 : Activation Function, optimizer, epochs, batch size 등


In [ ]:
!pip install pandas

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adagrad, Adadelta, RMSprop, Adam
import pandas as pd
import numpy as np
import os

In [ ]:
# 데이터 load
dataset = pd.read_csv(os.path.join(data_path, 'diabetes.csv'))

X = dataset.iloc[:, 0:8].values
y = dataset.iloc[:, 8].values

In [ ]:
# 수정 x
def configure_tf_determinism():
    # TensorFlow의 병렬성 감소
    tf.config.experimental.enable_op_determinism()

In [ ]:
# 실험 결과 재현을 위한 seed 고정
SEED = 4
os.environ['PYTHONHASHSEED'] = str(SEED)
os.environ['TF_DETERMINISTIC_OPS'] = '1'
tf.random.set_seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
initializer = tf.keras.initializers.GlorotUniform(seed=SEED)

In [ ]:
# 모델 생성
def build_model():
    model = Sequential()
    # 수정 가능
    # 활성화함수: relu 사용
    model.add(tf.keras.layers.Dense(12, input_dim=8, activation='relu', kernel_initializer=initializer))
    model.add(tf.keras.layers.Dense(8, activation='relu', kernel_initializer=initializer))

    # 수정 x
    model.add(tf.keras.layers.Dense(1, activation='sigmoid', kernel_initializer=initializer))

    # 옵티마이저: Adam으로 수정
    model.compile(loss='binary_crossentropy', optimizer='Adam', metrics=['accuracy'])

    return model

In [ ]:
# 수정 x
configure_tf_determinism()

In [ ]:
model = build_model()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
model.fit(X, y, epochs=10, batch_size=1)

Epoch 1/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.5768 - loss: 1.0543
Epoch 2/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.6432 - loss: 0.7325
Epoch 3/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.6745 - loss: 0.6736
Epoch 4/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6810 - loss: 0.6419
Epoch 5/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6797 - loss: 0.6282
Epoch 6/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6901 - loss: 0.6191
Epoch 7/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6927 - loss: 0.6041
Epoch 8/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6927 - loss: 0.6002
Epoch 9/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7070 - loss: 0.5901
Epoch 10/10
768/768 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7070 - loss: 0.5870


In [ ]:
_, accuracy = model.evaluate(X, y)

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7344 - loss: 0.5472  


In [ ]:
print('정확도: %.2f' % (accuracy*100))

정확도: 73.44
